app/src/androidTest/java/com/google/jetpackcamera/UiTestUtil.kt

```diff
@@ -76,3 +87,98 @@ fun ComposeTestRule.getCurrentLensFacing(): LensFacing {
         }
     }
 }
+
+/**
+ * Rule that you to specify test methods that will have permissions granted prior to starting
+ *
+ * @param permissions the permissions to be granted
+ * @param targetTestNames the names of the tests that this rule will apply to
+ */
+class IndividualTestGrantPermissionRule(
+    private vararg val permissions: String,
+    private val targetTestNames: List<String>
+) : TestRule {
+
+    override fun apply(base: Statement, description: Description): Statement {
+        return object : Statement() {
+            override fun evaluate() {
+                if (targetTestNames.contains(description.methodName)) {
+                    GrantPermissionRule.grant(*permissions)
+                } else {
+                    base.evaluate() // Run the test without granting permissions
+                }
+            }
+        }
+    }
+}
+
+// functions for interacting with system permission dialog
+@SdkSuppress(minSdkVersion = 30)
+fun askEveryTimeDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 30) {
+        val askPermission = findObjectByText(uiDevice = uiDevice, text = "Only this time")
+        askPermission!!.click()
+    }
+}
+
+/**
+ *  Clicks ALLOW option on an open permission dialog
+ */
+fun grantPermissionDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 23) {
+        val allowPermission = findObjectByText(
+            uiDevice = uiDevice,
+            text = when {
+                Build.VERSION.SDK_INT == 23 -> "Allow"
+                Build.VERSION.SDK_INT <= 28 -> "ALLOW"
+                Build.VERSION.SDK_INT == 29 -> "Allow only while using the app"
+                else -> "While using the app"
+            }
+        )
+        allowPermission?.click()
+    }
+}
+
+/**
+ * Clicks the DENY option on an open permission dialog
+ */
+fun denyPermissionDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 23) {
+        println("sdk version ${Build.VERSION.SDK_INT}")
+        val denyPermission = findObjectByText(
+            uiDevice = uiDevice,
+            text = when {
+                // todo fix this to be "Don't allow"... lol
+                // for some reason.... cannot find when the string has an apostrophe...
+                // i tried using \'...
+
+                Build.VERSION.SDK_INT >= 30 -> "Don"
```

Comment:
Similar to above, we should try to get these strings by res ID rather than relying on the text.

(btw, one way you might try escaping the apostrophe is with a unicode char: `Don\u2019t allow`)